### PR TITLE
perf(contexts) Promote span_id and trace_id contexts on the transactions storage.

### PR DESCRIPTION
**Pull request by fpacifici**
`trace_id` and `span_id` cannot be promoted like any other tag or context since the promoted columns are not strings.
Moreover even casting the result into string is not enough because the trace_id column is a proper UUID while the string in the contexts column does not have hyphens. The span_id is stored as an integer instead while the context contains a hex encoded string.

This requires a dedicated processor that promotes these columns when they are inside conditions and ignore them in the SELECT clause or group by so that the client still sees the expected format.

This is the first step. It just takes care of conditions. 
We can follow up by promoting the column everywhere and converting them back to what the user expects in select/groupby. This way we would not be loading the contexts column at all when possible.

### CODE REVIEW

File: snuba/query/processors/typed_context_promoter.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+
+class HexIntContextType(ContextType):
+    def _build_literal(self, query_value: str) -> LiteralExpr:
+        return LiteralExpr(None, int(query_value, 16))
+
+
+class PromotionSpec(NamedTuple):
+    context_name: str
+    promtoed_column_name: str
```
Comment (evanh) [inline snuba/query/processors/typed_context_promoter.py:72]:
```suggestion
    promoted_column_name: str
```

File: snuba/query/processors/typed_context_promoter.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+class TypedContextPromoter(QueryProcessor):
+    """
+    Promotes conditions on contexts that are not of string types.
+
+    This acts only on the WHERE clause and does not replace anything
+    in the select or group by clauses. Such improvements can be done
+    to further reduce the amount of data fetched.
+
+    The conditions it promtoes are of the form:
+    `arrayValue(contexts.values, indexOf(contexts.key, 'NAME')) = 'string_uuid'`
```
Comment (evanh) [inline snuba/query/processors/typed_context_promoter.py:85]:
```suggestion
    `arrayElement(contexts.values, indexOf(contexts.key, 'NAME')) = 'string_uuid'`
```

File: tests/query/processors/query_builders.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    return build_mapping_expr(
+        alias=f"{column}[{key}]",
+        table_name=None,
+        col_name=column,
+        mapping_key=Literal(None, key),
+    )
+
+
+def nested_condition(
+    column_name: str, operator: str, key: str, val: str,
```
Comment (evanh) [inline tests/query/processors/query_builders.py:43]: Nit: this is a weird order for these arguments. I would expect the operator to be between the LHS and RHS params.

File: snuba/query/processors/typed_context_promoter.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    `promoted_column = the_proper_representation_of_string_uuid`
+    """
+
+    def __init__(self, mapping_column: str, specs: Set[PromotionSpec]) -> None:
+        self.__mapping_column = mapping_column
+
+        # Generates the patterns to match upfront.
+        self.__patterns = {
+            spec.context_name: condition_pattern(
+                {ConditionFunctions.EQ},
```
Comment (evanh) [inline snuba/query/processors/typed_context_promoter.py:96]: Do you plan to add support for the IN operator later?

Comment (fpacifici) [reply snuba/query/processors/typed_context_promoter.py:96]: Yes, It could be a good idea.

**Review by evanh — APPROVED** [review]

**Pull request merged**